internal/controller/provisioner/objects.go

```diff
@@ -895,6 +1015,216 @@ func (p *NginxProvisioner) buildImage(nProxyCfg *graph.EffectiveNginxProxy) (str
+	// Add WAF containers
+	wafContainers := p.buildWAFContainers(nProxyCfg)
+	containers = append(containers, wafContainers...)
+
+	// Add WAF volume mounts to NGINX container
+	nginxContainer := containers[0]
+	nginxContainer.VolumeMounts = append(nginxContainer.VolumeMounts, buildNginxWAFVolumeMounts()...)
+	containers[0] = nginxContainer
+
+	// Add WAF volumes
+	wafVolumes := buildWAFSharedVolumes()
+	volumes = append(volumes, wafVolumes...)
+
+	return containers, volumes
+}
+
+// buildWAFSharedVolumes creates the required shared volumes for WAF containers.
+func buildWAFSharedVolumes() []corev1.Volume {
+	return []corev1.Volume{
+		{
+			Name:         appProtectBundlesVolumeName,
+			VolumeSource: emptyDirVolumeSource,
+		},
+		{
+			Name:         appProtectConfigVolumeName,
+			VolumeSource: emptyDirVolumeSource,
+		},
+		{
+			Name:         appProtectBdConfigVolumeName,
+			VolumeSource: emptyDirVolumeSource,
+		},
+	}
+}
+
+// buildNginxWAFVolumeMounts creates the required volume mounts for NGINX container when WAF is enabled.
+func buildNginxWAFVolumeMounts() []corev1.VolumeMount {
+	return []corev1.VolumeMount{
+		{
+			Name:      appProtectBundlesVolumeName,
+			MountPath: "/etc/app_protect/bundles",
+		},
+		{
+			Name:      appProtectConfigVolumeName,
+			MountPath: "/opt/app_protect/config",
+		},
+		{
+			Name:      appProtectBdConfigVolumeName,
+			MountPath: "/opt/app_protect/bd_config",
+		},
+	}
+}
+
+// buildWAFContainers creates the WAF enforcer and config manager containers.
+func (p *NginxProvisioner) buildWAFContainers(nProxyCfg *graph.EffectiveNginxProxy) []corev1.Container {
+	var containers []corev1.Container
+	var wafContainersCfg *ngfAPIv1alpha2.WAFContainerSpec
+
+	// Get WAF container configuration
+	if nProxyCfg != nil && nProxyCfg.Kubernetes != nil {
+		if nProxyCfg.Kubernetes.Deployment != nil {
+			wafContainersCfg = nProxyCfg.Kubernetes.Deployment.WAFContainers
+		} else if nProxyCfg.Kubernetes.DaemonSet != nil {
+			wafContainersCfg = nProxyCfg.Kubernetes.DaemonSet.WAFContainers
+		}
+	}
+
+	// Build WAF Enforcer container
+	enforcerContainer := p.buildWAFEnforcerContainer(wafContainersCfg)
+	containers = append(containers, enforcerContainer)
+
+	// Build WAF Config Manager container
+	configMgrContainer := p.buildWAFConfigManagerContainer(wafContainersCfg)
+	containers = append(containers, configMgrContainer)
+
+	return containers
+}
+
+// buildWAFEnforcerContainer creates the WAF enforcer container.
+func (p *NginxProvisioner) buildWAFEnforcerContainer(
+	wafContainersCfg *ngfAPIv1alpha2.WAFContainerSpec,
+) corev1.Container {
+	image := p.buildWAFImage(
+		defaultWAFEnforcerImagePath,
+		defaultWAFImageTag,
+		wafContainersCfg,
+		"enforcer",
+	)
+
+	container := corev1.Container{
+		Name:            "waf-enforcer",
+		Image:           image,
+		ImagePullPolicy: defaultImagePullPolicy,
+		SecurityContext: &corev1.SecurityContext{
+			RunAsUser: helpers.GetPointer[int64](101),
```

Comment:
Is there more that we should be adding to this SecurityContext? Drop capabilities, read only, etc?

Reply:
I took this from the NIC implementation, but I'll see if I can tweak it more.

Reply:
Resolving this in favour of below thread